### PR TITLE
Fig logging with log_gpu_memory='min_max'

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/9010

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
I think we should consider reverting the changes in https://github.com/PyTorchLightning/pytorch-lightning/pull/8174 and exploring other options.

See the open review comments in the PR.

Additionally, calling `lightning_module.log` from the `LoggerConnector` seems like a very bad design pattern which we should avoid. As that function is meant to be used by users.

Since it's quite possible that `log_gpu_memory` gets removed in favor of the `GPUStatsMonitor` callback, I think we should instead `lightning_module.log` within that callback. cc @ananthsub

These comments are not strictly related to the changes you did here @SkafteNicki. Just commenting here because reverting it would make this patch unnecessary.

**Comment by ananthsub** [comment]
@carmocca I agree - I don't think this logic needs to sit in the trainer this way. calling the lightning module's log from within the trainer, which under the hood actually calls back into the trainer is too hard to keep track of :( 

`mode=min_max` is specific to nvidia-smi. There's another open issue around using torch.cuda.memory directly for logging, especially for pytorch 1.8 and above: https://github.com/PyTorchLightning/pytorch-lightning/issues/8780

The more places that rely on `mode=min_max` the harder this will be to transition, so I'm in favor of corralling this into one place

cc @edward-io

**Review by awaelchli — APPROVED** [review]
@carmocca @ananthsub while I agree #8174 was not optimal, it broke what is being fixed here so I think we should still go ahead with Nicki's fix. And we have the adjustments to the test as well.

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 1]
```diff
@@ -645,15 +645,16 @@ def training_step(self, batch, batch_idx):
...
+def test_log_gpu_memory_without_logging_on_step(tmpdir, log_gpu_memory):
 
     model = BoringModel()
     trainer = Trainer(
         default_root_dir=tmpdir,
         max_epochs=1,
         limit_train_batches=1,
         limit_val_batches=0,
-        log_gpu_memory="all",
+        log_gpu_memory=log_gpu_memory,
```
Comment (awaelchli) [inline tests/trainer/logging_/test_train_loop_logging.py:657]: for min_max, need to adjust the assert below since the key will be different .)

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
> @carmocca @ananthsub while I agree #8174 was not optimal, it broke what is being fixed here so I think we should still go ahead with Nicki's fix. And we have the adjustments to the test as well.

I agree, we have to fix forward here. I filed https://github.com/PyTorchLightning/pytorch-lightning/issues/9032 to see how we can better formalize this

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
